homeassistant/components/vodafone_station/coordinator.py

```diff
@@ -125,6 +130,66 @@ async def _async_update_data(self) -> UpdateCoordinatorDataType:
         }
         return UpdateCoordinatorDataType(data_devices, data_sensors)
 
+    async def cleanup_device_tracker_entities(self) -> None:
+        """123."""
+        entity_reg: er.EntityRegistry = er.async_get(self.hass)
+
+        assert self.config_entry
+
+        ha_entity_reg_list: list[er.RegistryEntry] = er.async_entries_for_config_entry(
+            entity_reg, self.config_entry.entry_id
+        )
+        entities_removed: bool = False
+
+        device_hosts_macs = set()
+        device_hosts_names = set()
+        for mac, device_info in self.data.devices.items():
+            device_hosts_macs.add(mac)
+            device_hosts_names.add(device_info.device.name)
+
+        for entry in ha_entity_reg_list:
+            if entry.original_name is None:
+                continue
+            entry_name = entry.name or entry.original_name
+            entry_host = entry_name.split(" ")[0]
+            entry_mac = entry.unique_id.split("_")[0]
+
+            if entry.platform != DEVICE_TRACKER_DOMAIN or (
+                entry_mac in device_hosts_macs and entry_host in device_hosts_names
+            ):
+                _LOGGER.debug(
+                    "Skipping entity %s [mac=%s, host=%s]",
+                    entry_name,
+                    entry_mac,
+                    entry_host,
+                )
+                continue
+            _LOGGER.info("Removing entity: %s", entry_name)
+            entity_reg.async_remove(entry.entity_id)
+            entities_removed = True
+
+        if entities_removed:
+            self._async_remove_empty_devices(entity_reg, self.config_entry)
+
+    @callback
+    def _async_remove_empty_devices(
+        self, entity_reg: er.EntityRegistry, config_entry: ConfigEntry
+    ) -> None:
+        """Remove devices with no entities."""
+
+        device_reg = dr.async_get(self.hass)
+        device_list = dr.async_entries_for_config_entry(
+            device_reg, config_entry.entry_id
+        )
+        for device_entry in device_list:
+            if not er.async_entries_for_device(
+                entity_reg,
+                device_entry.id,
+                include_disabled_entities=True,
+            ):
+                _LOGGER.info("Removing device: %s", device_entry.name)
+                device_reg.async_remove_device(device_entry.id)
```

Comment:
```suggestion
                device_reg.async_update_device(
                    device_entry.id, remove_config_entry_id=self.config_entry.entry_id
                )
```

we should not remove a device entity entry, but remove the current config entry from it - if this was the last linkes config entry, the device registry will remove the device entry by its own (_see [here](https://developers.home-assistant.io/docs/device_registry_index#removing-devices) in the dev-docs_)
Background is, that this device entry can still be linked to another config entry.

Reply:
I've been working on this feature for the Plugwise integration, I've noticed when using the above proposed function, the connected entity-entries are automatically removed as well. So no need to separately call ` entity_reg.async_remove()` for all the present device-entities :)

Reply:
@bouwew please, to not overwhelm this PR, let's move this discussion to "your" PR were you're going to implement this, thx 👍

Reply:
@mib1185 please see https://github.com/plugwise/plugwise-beta/pull/627#issuecomment-2105633168